rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+    input.
+
+Next, we will show the code for the API as well as sample usage examples.
+
+#### Replicator class
+
+```python
+Nest = ...  # Type supported by `tf.nest`.
+T = TypeVar("T", Nest[tf.Tensor])  # A generic `Tensor` nest.
+U = TypeVar("U", Nest[tf.Tensor])  # Another generic `Tensor` nest.
+
+InputGenerator = Callable[[], T]
+InputFn = Callable[[InputContext], Union[tf.data.Dataset, InputGenerator[T]]]
+
+class Replicator(object):
+
+  def __init__(self, distribution: DistributionStrategy):
+    """Creates a `Replicator` with the given distribution strategy."""
+
+  def initialize(self) -> List[tf.Operation]:
+    """Initialization to be performed before running any computations.
+
+    In eager mode, it performs any necessary initialization.
+    In graph mode, it creates initialization ops and returns them.
+
+    In graph mode, the returned list will include the initialization ops for any
+    `tf.data.Dataset` iterators created by preceding calls to `prepare_input`.
+
+    Returns:
+      In eager mode, returns an empty list.
+      In graph mode, returns the list of ops to execute.
+    """
+
+  def finalize(self) -> List[tf.Operation]:
+    """Finalization to be performed after the end of all computations.
+
+    In eager mode, it performs any necessary finalization.
+    In graph mode, it creates finalization ops and returns them.
+
+    Returns:
+      In eager mode, returns an empty list.
+      In graph mode, returns the list of ops to execute.
+    """
+
+  def scope(self) -> ContextManager:
+    """Returns a context for model / `Variable` creation.
+
+    Models used within `run` must be created in this scope, so that `Variable`s
+    can be correctly replicated and / or placed on the correct devices.
+
+    N.B. This context is automatically applied within `run`.
+    """
+
+  def prepare_input(
+      self,
+      fn: InputFn[T],
+      replication_mode: InputReplicationMode = InputReplicationMode.PER_WORKER,
+      prefetch_on_device: Optional[bool] = None,
+      enforce_ordering: bool = False) -> InputIterator[T]:
+    """Prepares the input data.
+
+    Depending on the input replication mode, the input function may be called one time
+    or several. We call the result of each call an “input pipeline”. Disjoint subsets of the compute
+    replicas will draw their input data from each input pipeline. See `InputReplicationMode` for
+    more details.
+
+    If `fn` returns an instance of `tf.data.Dataset`, a corresponding `tf.data.Iterator` will be
+    created. Each replica taking its input from this pipeline will execute `Iterator.get_next()`.
+    The `tf.data.Iterator`(s) can be re-initialized by calling the `reinitialize` on the returned
+    `InputIterator` object.
+
+    If `fn` returns a Python callable (e.g. `FIFOQueue.dequeue`), that callable will be executed on
+    each replica taking its input from this pipeline. `InputIterator.reinitialize` will have no effect on
+    this pipeline.
```

Comment:
Returning a callable/Queue seems strange in the 2.0 world. Do we need to support this?